hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -842,17 +886,168 @@ case class MatrixAggregateColsByKey(child: MatrixIR, aggIR: IR) extends MatrixIR
+                    ir.I32(nAggs)),
+                  i),
+                args, aggSig)
+            case _ =>
+              ir.MapIR(rewrite)(x)
+          }
+        }
+
+        ir.ArrayFor(
+          ir.ArrayRange(ir.I32(0), ir.I32(oldNCols), ir.I32(1)),
+          colIdx,
+          ir.Let("sa", ir.ArrayRef(ir.Ref("colValues", oldColsType), ir.Ref(colIdx, TInt32())),
+            rewrite(sequenceIR)
+          ))
+      }
+    )
+
+    val (rTypCol, makeAnnotateCol) = ir.Compile[Long, Long, Long](
+      "AGGR", aggResultTypeCol,
+      "global", oldGlobalsType,
+      postAggIRCol
+    )
+    assert(coerce[TStruct](rTypCol) == newColValueType)
+
+    val nAggsCol = rvAggsColTemplate.length
+    val rvAggsCol = new Array[RegionValueAggregator](nAggsCol * nKeys)
+    var i = 0
+    while (i < nKeys) {
+      var j = 0
+      while (j < nAggsCol) {
+        rvAggsCol(i * nAggsCol + j) = rvAggsColTemplate(j).newInstance()
+        j += 1
+      }
+      i += 1
+    }
+
+    val rvb = new RegionValueBuilder()
+    
+    val newColValues = Region.scoped { region =>
+      rvb.set(region)
+
+      val globals = {
+        rvb.start(oldGlobalsType)
+        rvb.addAnnotation(oldGlobalsType, oldGlobalsBc.value)
+        rvb.end()
+      }
+
+      val cols = {
+        rvb.start(oldColsType)
+        rvb.addAnnotation(oldColsType, oldColValuesBc.value)
+        rvb.end()
+      }
+
+      val colIndices = {
+        rvb.start(newColumnIndicesType)
+        rvb.startArray(newColumnIndices.length)
+        var i = 0
+        while (i < newColumnIndices.length) {
+          rvb.addInt(newColumnIndices(i))
+          i += 1
+        }
+        rvb.endArray()
+        rvb.end()
+      }
+
+      makeInitCol(0)(region, rvAggsCol, globals, false)
+      makeSeqCol(0)(region, rvAggsCol, globals, false, cols, false, colIndices, false)
+
+      val annotateF = makeAnnotateCol(0)
+
+      val mapF = (a: Annotation, i: Int) => {
+        val aggResults = {
+          rvb.start(aggResultTypeCol)
+          rvb.startStruct()
+          var j = 0
+          while (j < nAggsCol) {
+            rvAggsCol(i * nAggsCol + j).result(rvb)
+            j += 1
+          }
+          rvb.endStruct()
+          rvb.end()
+        }
+
+        val colKeys = {
+          rvb.start(newColKeyType)
+          rvb.addAnnotation(newColKeyType, a)
+          rvb.end()
+        }
+
+        val colValues = annotateF(region, aggResults, false, globals, false)
+
+        val result = {
+          rvb.start(newColType)
+          rvb.startStruct()
+          rvb.addAllFields(newColKeyType, region, colKeys)
+          rvb.addAllFields(newColValueType, region, colValues)
+          rvb.endStruct()
+          rvb.end()
+        }
+
+        SafeRow(typ.colType.physicalType, region, result)
+      }
+
+      BroadcastIndexedSeq(keys.zipWithIndex.map { case (k, idx) => mapF(k, idx) }, TArray(typ.colType), hc.sc)
```

Comment:
you can just invert this. Instead of using this mapF structure, just directly map `keys` above.